harness/config/commands.yml

```
@@ -112,6 +112,20 @@ command('console'):
     #!bash(workspace:/)|@
     passthru $COMPOSE_BIN exec -u "${CODE_OWNER}" console bash
 
+command('shell [-u=<user>|--user=<user>] <service>', 'shell'):
+  description: Open a default (root) shell on any running container
+  env:
+    COMPOSE_BIN: = @('docker.compose.bin')
+    ARG_SERVICE: = input.argument('service')
+    OPT_USER: "= input.option('user') ?: 'root'"
+  exec: |
+    #!bash(workspace:/)|=
+    if echo "$OPT_USER" | grep -qv '^[a-z][-a-z0-9_]*$'; then
+      echo "Username pattern not allowed"
+      exit 1
+    fi
+    passthru $COMPOSE_BIN exec -u $OPT_USER $ARG_SERVICE $($COMPOSE_BIN exec -T $ARG_SERVICE cat /etc/passwd | grep "^$OPT_USER:" | sed -E 's/^.*:([^:]+)$/\1/')
+
```

Avoid default to root as that's just your preference, instead use the docker image's default user by default

```suggestion
command('shell [-u=<user>|--user=<user>] <service>', 'shell'):
  description: Open a default (root) shell on any running container
  env:
    COMPOSE_BIN: = @('docker.compose.bin')
    ARG_SERVICE: = input.argument('service')
    OPT_USER: "= input.option('user')"
  exec: |
    #!bash(workspace:/)|=
    if [ -z "$OPT_USER" ] || echo "$OPT_USER" | grep -qv '^[a-z][-a-z0-9_]*$'; then
      echo "Username pattern not allowed"
      exit 1
    fi
    passthru $COMPOSE_BIN exec ${OPT_USER:+-u "$OPT_USER"} "$ARG_SERVICE" "$($COMPOSE_BIN exec -T $ARG_SERVICE cat /etc/passwd | grep "^$OPT_USER:" | sed -E 's/^.*:([^:]+)$/\1/')"

```

---

also wondering if can use `getent passwd "$OPT_USER"` on all linux distros

---

Well as you remember the original intent was `root` but I see the point. `ws root` is free to use again as alias anyway 😅 
I didn't know about `getent`, from what I read that should be usable on any distro so I'll change to that.

---

one benefit of getent is that it removes the grep need, which might be different behaviour on MacOS to linux or alpine linux as to ^ (the pipes run on host), and also removes the need for username pattern checks (if that was due to security)

---

Running the pipes on host was actually accidental so I changed that anyway. From what I read, using `^` as anchor should never require extended regex, only as negation (that's where I'm not certain whether all `-E`s work the same). Another benefit of using `getent` is that I got to know it 😄

---

ok another TUL (Today you learned) about `getent`, `getent hosts google.com` = dns resolution